docs(compactor2): Update compactor2 config parameter documentation

alamb opened the pull request:
my secret motive is to reduce the number of these knobs. The first step of this is more fully understanding what the current knobs do and I wanted to encode my learnings for others (and an improved CLI experience)

alamb commented on clap_blocks/src/compactor2.rs:
```diff
@@ -98,6 +107,7 @@ pub struct Compactor2Config {
     pub max_desired_file_size_bytes: u64,
 
     /// Percentage of desired max file size.
```
I am not sure how this parameter is different than `split_percentage` -- I am clarifying on slack

NGA-TRAN replied:
1. - When compacted output is  smaller than size S, the data is written into one file
2. - If the compacted output is between S and M, we split them into 2 files. The first one is split_percentage and the second one is (100% - split_percentage )
3. - If compacted size > M, we split them into many files, each size around M except the last one which can be smaller than M.

- The percentage_max_file_size is used to determine S.  S = percentage_max_file_size * M
- M = max_desired_file_size_bytes

Values of these 3 flags are from our experiments on customer data before to reduce re-compacting data. If we can have fewer flags that do the similar job, it should shoot for it

alamb replied:
Documented split percentage here: https://github.com/influxdata/influxdb_iox/pull/7026

NGA-TRAN commented on clap_blocks/src/compactor2.rs:
```diff
@@ -65,7 +70,8 @@ pub struct Compactor2Config {
 
-    /// Number of threads to use for the compactor query execution, compaction and persistence.
+    /// Number of threads to use for the compactor query execution,
+    /// compaction and persistence.
```
From talking with @crepererum the other day, this value should be set to `the number of core - 1`. I think we should add this into the comment

alamb replied:
I think it would make sense to make that change in a separate PR (and actually default to `number of cores - 1`)

We can follow the model of

https://github.com/influxdata/influxdb_iox/blob/b785f751b354c65d2e0548972f9e94b50e118612/clap_blocks/src/querier.rs#L50-L58


@Joe-Blount  is that something that would be interesting to try and do? It might be a good way to get more Rust exposure

Joe-Blount replied:
Sure!

alamb replied:
I am happy to file a follow on ticket to track this work if you like, or if you are tracking items in some other way that is cool too. Just let me know @Joe-Blount

Joe-Blount replied:
@alamb I think it is as simple as this: https://github.com/influxdata/influxdb_iox/pull/7025

NGA-TRAN commented on clap_blocks/src/compactor2.rs:
```diff
@@ -74,10 +80,12 @@ pub struct Compactor2Config {
-    /// during execution, they will error with "ResourcesExhausted".
+    /// If compaction plans attempt to allocate more than this many
+    /// bytes during execution, they will error with
+    /// "ResourcesExhausted".
```
Great to have this clear description and expectation

NGA-TRAN commented on clap_blocks/src/compactor2.rs:
```diff
@@ -169,7 +179,14 @@ pub struct Compactor2Config {
+    /// Maximum number of files that the compactor will try and
+    /// compact in a single plan.
+    ///
+    /// The higher this setting is the fewer compactor plans are run
```
> The higher this setting is the fewer compactor plans are run

I do not thing this is accurate. This flag is for the limit of number of files per plan. If there are more, we will split them and make sure number of files per plan < this value.

1. Whenever a partition has new L0 files, we will compact the with overlapped L1.
2. Whenever the total size of L1 files > max_desired_file_size, we will compact them with overlapped L2s
3. Whenever number of L1 files > min_num_l1_files_to_compact, we will compact them with overlapped L2s

alamb replied:
> This flag is for the limit of number of files per plan. If there are more, we will split them and make sure number of files per plan < this value.

Your explanation is the internal description of *how* the algorithm uses this number.

I am trying to describe this knob in terms of something a user/operator can observe or know about (which is not file levels or their overlap in my opinion)

I think my description is accurate
* if you set this number higher, the "compact many small files" mode will happen less often and thus run few plans. 
* If you set it lower there are more files that may be compacted in any L0->L1 round (and likewise L1 -> L2 rounds) and thus each plan may require more memory

NGA-TRAN replied:
While I agree we do not need to describe how, I think saying setting value of this flag high will make compact run less but each time require more memories is not that it means because we do not do it.

I think we may not want to mention the run less/more because it does not for that and we do not do it. We may just need to mention this is to limit the memory needed of running each plan maybe 🤔

alamb replied:
> , I think saying setting value of this flag high will make compact run less but each time require more memories is not that it means because we do not do it.

How about "*may* require more memory each time"? 

If the description is only "lowering this knob will lower the memory used by the compactor" -- then the obvious question people will ask is "why not turn it all the way down"? There must be some tradeoff for turning it down (which I think is more plans can potentially be run)

alamb replied:
> I think we may not want to mention the run less/more because it does not for that and we do not do it.

@NGA-TRAN  sorry, I  must be missing something . Here is an example showing that if `with_max_num_files_per_plan` is set higher there are fewer plans but those plans use more memory:  https://github.com/influxdata/influxdb_iox/pull/7024

Why are you saying this is not accurate?

NGA-TRAN replied:
Hah, I see it.
Let think of a use case where the compactor is behind. When it compacts a partition of 100 L0 files and the knob is set to 20. This means we have to split 100 files into 5 groups each 20 files. Compact them into 5 L0 files and then compact those 5 L0 files into L1s.

So I guess your `The higher this setting is the fewer compactor plans are run` is accurate  🙂

NGA-TRAN commented on clap_blocks/src/compactor2.rs:
```diff
@@ -178,8 +195,13 @@ pub struct Compactor2Config {
-    /// self-protection mechanism.
+    /// Maximum input bytes (in parquet) per partition.
+    ///
+    /// If the sum of the sizes of all parquet files in a partition is greater than
+    /// this value, the compactor will not try to compact this partition.
```
```suggestion
    /// If the sum of the sizes of `files_to_compact` is greater than
    /// this value, the compactor will not try to compact this partition.
```

and I think you want to define `files_to_compact` but I know you know what it is. Number of `files_to_compact` is usually a lot smaller than number of files in the partition. That is why we do not hit this case often after we turn `TargetLevel` on

alamb replied:
Since `files_to_compact` is some part of the internal calculation of the compactor and will not be easy for users to compute. 

I will try and clarify this wording to make it clear it relates to the worst case (when all files are being compacted together)

alamb replied:
Clarified in f1820732e

NGA-TRAN reviewed (approved):
There are a few definitions need to modify but I provided their meaning in the detailed comments

alamb commented:
I think I have addressed all feedback on this PR so I am going to merge it in and I'll make a follow on one trying to explain `split_percentage` and `compaction-percentage-max-file_size`
